cinnabar/femap.py

```diff
@@ -0,0 +1,235 @@
+
+    @classmethod
+    def from_csv(cls, filename, units: Optional[unit.Quantity] = None):
+        """Construct from legacy csv format"""
+        data = read_csv(filename, units=units)
+
+        # unpack data dictionary
+        fe = cls()
+        for r in data['Calculated']:
+            fe.add_measurement(r)
+        for r in data['Experimental'].values():
+            fe.add_measurement(r)
+
+        return fe
+
+    def add_measurement(self, measurement: Union[RelativeMeasurement, AbsoluteMeasurement]):
+        """Add new observation to FEMap, modifies the FEMap in-place
+
+        Any other attributes on the measurement are used as annotations
+
+        Raises
+        ------
+        ValueError : if bad type given
+        """
+        if isinstance(measurement, AbsoluteMeasurement):
+            # coerce to relative to simplify logic
+            meas_ = RelativeMeasurement(labelA='NULL',
+                                        labelB=measurement.label,
+                                        DDG=measurement.DG,
+                                        uncertainty=measurement.uncertainty,
+                                        computational=measurement.computational)
+        elif isinstance(measurement, RelativeMeasurement):
+            meas_ = measurement
+        else:
+            raise ValueError()
+
+        # slurp out tasty data, anything but labels
+        d = dict(meas_)
+        d.pop('labelA', None)
+        d.pop('labelB', None)
+        d.pop('label', None)
+
+        self.graph.add_edge(meas_.labelA, meas_.labelB, **d)
+
+    @property
+    def n_measurements(self) -> int:
+        """Total number of both experimental and computational measurements"""
+        return len(self.graph.edges)
+
+    @property
+    def n_ligands(self) -> int:
+        """Total number of unique ligands"""
+        # must ignore NULL sentinel node
+        return len(self.graph.nodes - {'NULL'})
+
+    @property
+    def degree(self) -> float:
+        """Average degree of computational nodes"""
+        return self.n_edges / self.n_ligands
+
+    @property
+    def n_edges(self) -> int:
+        """Number of computational edges"""
+        return sum(1 for _, _, d in self.graph.edges(data=True) if d['computational'])
+
+    def check_weakly_connected(self) -> bool:
+        """Checks if all results in the graph are reachable from other results"""
+        # todo; cache
+        comp_graph = nx.MultiGraph()
+        for a, b, d in self.graph.edges(data=True):
+            if not d['computational']:
+                continue
+            comp_graph.add_edge(a, b)
+
+        return nx.is_connected(comp_graph)
+
+    def generate_absolute_values(self):
+        # TODO: Make this return a new Graph with computational nodes annotated with DG values
+
+        # TODO this could work if either relative or absolute expt values are provided
+        if self.check_weakly_connected():
+            graph = self.to_legacy_graph()
```

Comment:
I assume this is a temporary solution?

Reply:
yes.  as you've raised, things like MLE will need some rethinking in the light of absolute values, so for now `to_legacy_graph` works as before (and drops abs values)